src/passes/pass-utils.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2023 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_passes_pass_utils_h
+#define wasm_passes_pass_utils_h
+
+#include <unordered_set>
+
+#include <pass.h>
+#include <wasm.h>
+
+namespace wasm {
+
+namespace PassUtils {
+
+using FuncSet = std::unordered_set<Function*>;
+
+// A wrapper around a parallel pass that filters it to run run only on select
+// functions.
+struct FilteredPass : public Pass {
+  std::unique_ptr<Pass> create() override {
+    // Function-parallel passes get a new instance per function. Create a copy
+    // of the wrapped pass along with ourselves.
+    return std::make_unique<FilteredPass>(pass->create(), relevantFuncs);
+  }
+
+  FilteredPass(std::unique_ptr<Pass> pass, const FuncSet& relevantFuncs)
```

Comment:
This should be an rvalue reference so you don't have to move-construct the parameter just to move it again.
```suggestion
  FilteredPass(std::unique_ptr<Pass>&& pass, const FuncSet& relevantFuncs)
```

Reply:
How does that work? I can make this change, but I cannot remove either of the `std::move`s in this file (`use of deleted function` errors etc.), so I'm not sure how it helps.

Reply:
Before this change, the unique_ptr move constructor gets called twice. First, this `pass` parameter is move-constructed from whatever rvalue you pass into this `FilteredPass` constructor. Second, you move-construct the `pass` member from the `pass` parameter below in the initializer list with `pass(std::move(pass))`.

After this change, `pass(std::move(pass))` would move-construct the `pass` member from the `rvalue` reference you pass into the `FilteredPass` constructor without move-constructing the `pass` parameter as a separate intermediate value in the middle.

I'm not sure if it makes any difference after optimizations, but in principle this change results in one less call to the unique_ptr move constructor.

Reply:
Interesting, thanks.

I have very little idea of how the optimizer handles this stuff, but sounds like it can help. Added in the last push.